pallets/ddc-verification/src/lib.rs

```diff
@@ -100,19 +214,112 @@ pub mod pallet {
 				verification_key
 					.clone()
 					.try_into()
 					.map_err(|_| Error::<T>::BadVerificationKey)?;
 
-			let receipt_params =
-				ReceiptParams { verification_key: bounded_verification_key, merkle_root_hash };
+			VerificationKey::<T>::put(bounded_verification_key);
+			Self::deposit_event(Event::<T>::VerificationKeyStored { verification_key });
 
-			ActiveBillingReports::<T>::insert(cluster_id, era, receipt_params);
+			Ok(())
+		}
+
+		#[pallet::call_index(2)]
+		#[pallet::weight(T::WeightInfo::create_billing_reports())]
+		pub fn set_validate_payout_batch(
+			origin: OriginFor<T>,
+			cluster_id: ClusterId,
+			era: DdcEra,
+			payout_data: PayoutData<T::Hash>,
+		) -> DispatchResult {
+			let who = ensure_signed(origin)?;
+			let account_bytes: [u8; 32] = Self::account_to_bytes(&who)?;
+			let who: T::AuthorityIdParameter =
+				sp_application_crypto::sr25519::Public::from_raw(account_bytes).into();
+			let who: T::AuthorityId = who.into();
+			let authorities = <ValidatorSet<T>>::get().unwrap();
+
+			ensure!(authorities.contains(&who.clone()), Error::<T>::NotAValidator);
+
+			ensure!(
+				!<PayoutValidators<T>>::get(cluster_id, era).contains(&who.clone()),
+				Error::<T>::AlreadySigned
+			);
+
+			<PayoutValidators<T>>::try_mutate(cluster_id, era, |validators| -> DispatchResult {
+				validators.push(who);
+				Ok(())
+			})?;
+
+			const MAJORITY: u8 = 67;
+			let p = Percent::from_percent(MAJORITY);
+			let threshold = p * authorities.len();
+
+			let signed_validators = <PayoutValidators<T>>::get(cluster_id, era);
+
+			if threshold < signed_validators.len() {
+				PayoutBatch::<T>::insert(cluster_id, era, payout_data);
+				Self::deposit_event(Event::<T>::PayoutBatchCreated { cluster_id, era });
+			}
 
-			Self::deposit_event(Event::<T>::BillingReportCreated { cluster_id, era });
 			Ok(())
 		}
 	}
 }
+
+impl<T: Config> Pallet<T> {
+	// This function converts a 32 byte AccountId to its byte-array equivalent form.
+	fn account_to_bytes<AccountId>(account: &AccountId) -> Result<[u8; 32], DispatchError>
+	where
+		AccountId: Encode,
+	{
+		let account_vec = account.encode();
+		ensure!(account_vec.len() == 32, "AccountId must be 32 bytes.");
+		let mut bytes = [0u8; 32];
+		bytes.copy_from_slice(&account_vec);
+		Ok(bytes)
+	}
+}
+impl<T: Config> sp_application_crypto::BoundToRuntimeAppPublic for Pallet<T> {
+	type Public = T::AuthorityId;
+}
+
+impl<T: Config> OneSessionHandler<T::AccountId> for Pallet<T> {
+	type Key = T::AuthorityId;
+
+	fn on_genesis_session<'a, I: 'a>(validators: I)
+	where
+		I: Iterator<Item = (&'a T::AccountId, Self::Key)>,
+	{
+		let validators = validators.map(|(_, k)| k).collect::<Vec<_>>();
+
+		ValidatorSet::<T>::put(validators);
+	}
+
+	fn on_new_session<'a, I: 'a>(_changed: bool, validators: I, _queued_authorities: I)
+	where
+		I: Iterator<Item = (&'a T::AccountId, Self::Key)>,
+	{
+		let validators = validators.map(|(_, k)| k).collect::<Vec<_>>();
+		ValidatorSet::<T>::put(validators);
```

Comment:
shouldn't you clear the ValidatorSet before, as it contains previous session validators?